### ISSUE TITLE
jq add '\r' at the end of variable

### ISSUE THREAD

**Issue by cgaulay** [stedolan/jq#1854]
Hi all,

**Describe the bug**
When i use jq like this : 
`foo=$(jq -n --arg bar "$toto" $test )`
jq seems to add '\n' character at the end of $toto variable.

>  "foo": "a50921ca-5df8-4f66-b4e9-0f8a2584dcb0\r",

I don't understand why jq do that.

**Expected behavior**
put variable without new character '\r' 
> "foo":"a50921ca-5df8-4f66-b4e9-0f8a2584dcb0",

**Environment (please complete the following information):**
 - jq-1.5-1-a5b5cbe

**Comment by nicowilliams** [comment]
jq does no such thing.  Can you show what `$toto` is in your shell?  Do it like this:

```
$ printf '%s' "$toto" | od -c
```

**Comment by linquize** [comment]
Are you using Windows? jq on Windows uses `\r\n` as EOL.

**Comment by nicowilliams** [comment]
Even on Windows jq doesn't do this:

```
    C:\Users\nico\Downloads>"C:\Program Files\Git\bin\bash"

    nico@DESKTOP-NHQB0H9 MINGW64 ~/Downloads
    $ toto=foo

    nico@DESKTOP-NHQB0H9 MINGW64 ~/Downloads
    $ echo "$toto"|od -c
    0000000   f   o   o  \n
    0000004

    nico@DESKTOP-NHQB0H9 MINGW64 ~/Downloads
    $ ./jq-win64.exe -n --arg toto "$toto" '$toto'|od -c
    0000000   "   f   o   o   "  \r  \n
    0000007

    nico@DESKTOP-NHQB0H9 MINGW64 ~/Downloads
    $ ./jq-win64.exe -n --arg toto "$toto" '{$toto}'|od -c
    0000000   {  \r  \n           "   t   o   t   o   "   :       "   f   o
    0000020   o   "  \r  \n   }  \r  \n
    0000027

    nico@DESKTOP-NHQB0H9 MINGW64 ~/Downloads
--->$ ./jq-win64.exe -cn --arg toto "$toto" '{$toto}'|od -c
--->0000000   {   "   t   o   t   o   "   :   "   f   o   o   "   }  \r  \n
    0000020

    nico@DESKTOP-NHQB0H9 MINGW64 ~/Downloads
    $
```

As you can see, jq does add `\r` at the end of each thing it prints on `stdout`, but it does not add `\r` to strings in the middle of the things it prints to `stdout`.  The `\r` in @cgaulay's `$toto` was already there before jq even got started.  Also, we don't know what `$test` is...  Just to be sure, I checked, and `tojson` doesn't add `\r\n`, so I'm not sure what non-obvious thing `$test` might do to add a carriage return to `$toto`.

@cgaulay can you give us more to go on?  What is the minimal jq program and shell script you can write to reproduce this?

**Comment by doggy8088** [comment]
@nicowilliams I can reproduce this problem.  Please check my screenshot.

![image](https://user-images.githubusercontent.com/88981/54282290-a8462d00-45d6-11e9-8a0a-2feb0f77472b.png)


This problem is hard to see in the Command Prompt:

![image](https://user-images.githubusercontent.com/88981/54282355-cdd33680-45d6-11e9-8854-f0aee8d1678b.png)

If you open this `aa.txt` using Notepad, you will see this:

![image](https://user-images.githubusercontent.com/88981/54282466-0410b600-45d7-11e9-8fe9-fa925ecde670.png)

**Comment by nicowilliams** [comment]
@doggy8088 see #1855 -- this is an interaction between a shell that is pretending Windows is Unix, and a jq built for native Windows which therefore assumes lines end in `CRLF`.

I'm not sure what to do about that.

Options include:

 - add a command-line option to make `stdout` binary
 - build and release Cygwin etc. jq executables for git-bash and Cygwin
 - ???

**Comment by doggy8088** [comment]
@nicowilliams I'm using pure native Command Prompt.  Do you see my steps on not using WSL?

**Comment by nicowilliams** [comment]
@doggy8088 oops, I missed that.

**Comment by doggy8088** [comment]
@nicowilliams I think `jq` simply replace all `\n` to `\r\n`.  This is not right on Windows.  Some JSON data already contains `\r\n` in their string. So you might need to remove `\r` first to normalize the data then output to STDOUT if RAW_OUTPUT is set.

**Comment by linquize** [comment]
On windows,  Text mode fprintf()  , fputs()  convert `\n` to `\r\n`

**Comment by nicowilliams** [comment]
jq does not change newlines to CRLFs.  The Windows C run-time or kernel do.  The only thing we could do is reopen `stdout` in binary mode.

**Comment by doggy8088** [comment]
Does `jq` able to output in binary mode?

**Comment by nicowilliams** [comment]
@doggy8088 Not at this time.  Perhaps we could add an option for it.

**Comment by nicowilliams** [comment]
@cgaulay @doggy8088 Can you test [this version](https://ci.appveyor.com/api/buildjobs/9pc3vq0ih6t3oeok/artifacts/jq.exe) (or [32-bit](https://ci.appveyor.com/api/buildjobs/scuf0tj7ulc57t8q/artifacts/jq.exe)) of jq for Windows, adding a `-b` or `--binary` command-line option?

**Comment by linquize** [comment]
`-b` option ok
`--binary' not ok: `Unknown option --binary`

**Comment by nicowilliams** [comment]
Ok, thanks, i know what to do.  I'll push this soon.

**Comment by nicowilliams** [comment]
Fixed in master with 0dab2b1.
